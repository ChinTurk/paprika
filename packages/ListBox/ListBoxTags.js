import React, { useState, useEffect } from "react";
import { string, node } from "prop-types";
import ListBox from "./ListBox";
import Option from "./components/Option";
import Group from "./components/Group";
import Tags from "./components/Tags";

const propTypes = {
  children: node.isRequired,
  placeholder: string,
};

export default function ListBoxWithTags(props) {
  const [tagsSelected, setTagsSelected] = useState([]);

  function renderLabel(_props, state, set) {
    const handleDeleteTag = key => () => {
      const tagsSelectedClone = tagsSelected.slice();
      const index = tagsSelected.indexOf(key);
      tagsSelectedClone.splice(index, 1);
      setTagsSelected(tagsSelectedClone);
      set(_state => {
        const selectedOptionsClone = _state.selectedOptions.slice();
        const _index = _state.selectedOptions.indexOf(key);
        selectedOptionsClone.splice(_index, 1);

        return {
          ..._state,
          activeElement: 0,
          isPopoverOpen: true,
          selectedOptions: selectedOptionsClone,
        };
      });
    };

    useEffect(() => {
      if (state.selectedOptions.length) {
        setTagsSelected(state.selectedOptions);
      }
    }, [state.selectedOptions]);

    return (
      <Tags {..._props}>
        {state.selectedOptions.length &&
          state.selectedOptions.map(key => (
            <Tags.Option key={state.options[key].id} onDelete={handleDeleteTag(key)}>
              {state.options[key].content}
            </Tags.Option>
          ))}
      </Tags>
    );
  }

  return (
    <ListBox {...props} isMulti renderLabel={renderLabel}>
      {React.Children.map(props.children, (child, index) => {
        if (tagsSelected.length && tagsSelected.includes(index)) {
          return React.cloneElement(child, { ...child.props, isHidden: true });
        }
        return child;
      })}
    </ListBox>
  );
}

ListBoxWithTags.propTypes = propTypes;
ListBoxWithTags.Option = Option;
ListBoxWithTags.Group = Group;
