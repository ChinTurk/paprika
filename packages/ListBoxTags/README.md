<!-- autogenerated don't modify -->

# @paprika/listbox-tags - 0.0.1

## Description

ListBoxTags behave as a ListBox with the difference that the selected option stack on pills next to an search input

## Installation

```
yarn add @paprika/listbox-tags
```

or with npm:

```
npm install @paprika/listbox-tags
```

## Props

### ListBoxTags

| Prop     | Type   | required | default | Description |
| -------- | ------ | -------- | ------- | ----------- |
| children | custom | false    | -       |             |

<!-- autogenerated don't modify -->
<!-- content -->

<!-- eoContent -->

## Links

- [Storybook showcase](https://paprika.highbond.com/?path=/story/listbox-tags--showcase)
- [Github source code](https://github.com/acl-services/paprika/tree/master/packages/ListBoxTags/src)
- [Github create issue](https://github.com/acl-services/paprika/issues/new?label=[]&title=@paprika/listbox-tags%20[help]:%20your%20short%20description&body=%0A%23%20Help%20wanted%0A%0A%23%23%20Please%20write%20your%20question.%0A*A%20clear%20and%20concise%20description%20of%20what%20the%20question%20is*%0A%0A%23%23%20Additional%20context%0A*Add%20any%20other%20context%20or%20screenshots%20about%20your%20question%20here.*%0A)
- [ChangeLog](https://github.com/acl-services/paprika/tree/master/packages/ListBoxTags/CHANGELOG.md)
