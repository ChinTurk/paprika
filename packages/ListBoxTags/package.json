{
  "name": "@paprika/listbox-tags",
  "version": "0.0.1",
  "description": "ListBoxTags behave as a ListBox with the difference that the selected option stack on pills next to an search input",
  "author": "@paprika",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/acl-services/paprika.git",
    "directory": "packages/ListBoxTags"
  },
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@babel/runtime-corejs2": "^7.3.1",
    "@paprika/avatar": "^0.1.7",
    "@paprika/button": "^0.3.14",
    "@paprika/constants": "^0.1.3",
    "@paprika/helpers": "0.2.13",
    "@paprika/icon": "^0.3.10",
    "@paprika/input": "^0.3.10",
    "@paprika/l10n": "^0.4.11",
    "@paprika/listbox": "^0.9.14",
    "@paprika/pill": "^0.2.31",
    "@paprika/popover": "^0.3.35",
    "@paprika/raw-button": "^0.2.20",
    "@paprika/stylers": "^0.2.11",
    "@paprika/tokens": "^0.1.17",
    "fuse.js": "^6.4.1",
    "prop-types": "^15.7.2"
  },
  "peerDependencies": {
    "react": "^16.8.4",
    "react-dom": "^16.8.4",
    "styled-components": "^4.2.0"
  },
  "paprikaDocs": {
    "subComponents": []
  }
}
