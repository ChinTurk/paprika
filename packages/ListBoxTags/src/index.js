export { default, ListBoxTags2 } from "./ListBoxTags";
