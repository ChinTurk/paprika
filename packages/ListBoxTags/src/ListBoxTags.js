import React from "react";
import PropTypes from "prop-types";
import ListBox from "@paprika/listbox";
import Input from "@paprika/input";
import Fuse from "fuse.js";
import Pill, { Pills } from "./components/Pill";
import * as sc from "./ListBoxTags.styles";

const propTypes = {
  data: PropTypes.arrayOf(PropTypes.object),
  isOptionSelected: PropTypes.func,
  onChange: PropTypes.func,
  onChangeFilter: PropTypes.func,
  renderOption: PropTypes.func,
};
const defaultProps = {
  data: [],
  renderOption: null,
  onChangeFilter: () => {},
  onChange: () => {},
  isOptionSelected: () => {},
};

function applyFilter(searchTerm, options) {
  const fuze = new Fuse(options, { includeScore: true, includeMatches: true, keys: ["label"] });
  return fuze.search(searchTerm).map(result => result.item);
}

const renderTrigger = ({ onChangeFilter, refListBox }) => (...args) => {
  const [selected, options, currentSelected, attributes] = args;
  const { propsForTrigger, refTrigger, dispatch, types, handleKeyDown, handleKeyUp } = attributes;

  function handleClick(event) {
    // we don't want to close the popover if the user click enter or space to select an option
    if (event.key === "Enter" || event.key === " ") return;
    dispatch({ type: types.togglePopover });
  }

  function handleChange(event) {
    onChangeFilter(event.target.value);
    if (refListBox.current && "setActiveOptionIndex" in refListBox.current) {
      /**
       * this wait a render cycle and then force the active option index (key up key down position)
       * to reset with zero index position, so when the user search and the list is dynamically changing
       * the index is always 0 and the user always can either click enter or space, up or down, to get the most
       * convenient option.
       */
      requestAnimationFrame(() => {
        refListBox.current.setActiveOptionIndex();
      });
    }
  }

  return (
    <div ref={refTrigger} {...propsForTrigger()}>
      <div>
        {selected.map(index => {
          return <div>{options[index].label}</div>;
        })}
      </div>
      <Input onClick={handleClick} onKeyUp={handleKeyUp} onKeyDown={handleKeyDown} onChange={handleChange} />
    </div>
  );
};

export default function ListBoxTags(props) {
  const { data, renderOption, onChange, onChangeFilter, isOptionSelected } = props;
  const refListBox = React.useRef(null);

  return (
    <ListBox ref={refListBox} isMulti onChange={onChange}>
      <ListBox.Popover shouldKeepFocus />
      <ListBox.Trigger>{renderTrigger({ onChangeFilter, refListBox })}</ListBox.Trigger>
      {data.map((option, index) => {
        if (typeof option.label !== "string")
          throw new Error("data array prop must have all their object items with an label (:string) property");

        if (typeof option.isDivider !== "undefined") {
          return <ListBox.Divider {...option}>{option.label}</ListBox.Divider>;
        }

        return (
          <ListBox.Option {...option} isSelected={isOptionSelected(index, option)} key={option.label}>
            {typeof renderOption === "function" ? renderOption(option) : option.label}
          </ListBox.Option>
        );
      })}
    </ListBox>
  );
}

const renderTrigger2 = (...args) => {
  const [selected, options, currentSelected, attributes] = args;
  const { propsForTrigger, refTrigger, dispatch, types, handleKeyDown, handleKeyUp } = attributes;

  function handleClick(event) {
    // we don't want to close the popover if the user click enter or space to select an option
    if (event.key === "Enter" || event.key === " ") return;
    dispatch({ type: types.togglePopover });
  }

  return (
    <sc.Trigger
      ref={refTrigger}
      {...propsForTrigger()}
      onClick={handleClick}
      onKeyUp={handleKeyUp}
      onKeyDown={handleKeyDown}
    >
      <Pills>
        {selected.map(index => {
          return <Pill>{options[index].label}</Pill>;
        })}
        {selected.length ? null : <div>Open me</div>}
      </Pills>
    </sc.Trigger>
  );
};

export function ListBoxTags2({ children, moreProps }) {
  return (
    <ListBox isMulti {...moreProps}>
      <ListBox.Trigger>{renderTrigger2}</ListBox.Trigger>
      <ListBox.Filter />
      {children}
    </ListBox>
  );
}

ListBoxTags.propTypes = propTypes;
ListBoxTags.defaultProps = defaultProps;
ListBoxTags.filter = applyFilter;
