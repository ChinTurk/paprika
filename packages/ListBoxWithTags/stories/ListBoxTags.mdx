import { Meta, Props, Story } from "@storybook/addon-docs/blocks";
import ListBoxTags from "../src/";

# ListBoxTags
