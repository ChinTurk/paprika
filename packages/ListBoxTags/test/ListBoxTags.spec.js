import ListBoxTags from "../src";

describe("ListBoxTags", () => {
  it("should filter correctly the option", () => {
    const data = [
      { id: 1, label: "The Dog Jumps" },
      { id: 2, label: "The Fox Jumps" },
      { id: 3, label: "The Raccoon Jumps" },
    ];

    expect(ListBoxTags.filter("Dog", data)).toMatchSnapshot();
    expect(ListBoxTags.filter("Fox", data)).toMatchSnapshot();
    expect(ListBoxTags.filter("Jum", data)).toMatchSnapshot();
    expect(ListBoxTags.filter("abds", data)).toMatchSnapshot();
  });
});
