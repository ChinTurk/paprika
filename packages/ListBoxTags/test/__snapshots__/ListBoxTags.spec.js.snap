// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListBoxTags should filter correctly the option 1`] = `
Array [
  Object {
    "id": 1,
    "label": "The Dog Jumps",
  },
]
`;

exports[`ListBoxTags should filter correctly the option 2`] = `
Array [
  Object {
    "id": 2,
    "label": "The Fox Jumps",
  },
]
`;

exports[`ListBoxTags should filter correctly the option 3`] = `
Array [
  Object {
    "id": 1,
    "label": "The Dog Jumps",
  },
  Object {
    "id": 2,
    "label": "The Fox Jumps",
  },
  Object {
    "id": 3,
    "label": "The Raccoon Jumps",
  },
]
`;

exports[`ListBoxTags should filter correctly the option 4`] = `Array []`;
